core/src/test/scala/unit/kafka/zk/ZooKeeperTestHarness.scala

```diff
@@ -54,3 +58,40 @@ abstract class ZooKeeperTestHarness extends JUnitSuite with Logging {
     Configuration.setConfiguration(null)
   }
 }
+
+object ZooKeeperTestHarness {
+
+  /**
+   * Verify that a previous test that doesn't use ZooKeeperTestHarness hasn't left behind an unexpected thread
+   */
+  @BeforeClass
+  def setUpClass() {
+    verifyNoUnexpectedThreads()
+  }
+
+  /**
+   * Verify that tests from the current test class using ZooKeeperTestHarness hasn't left behind an unexpected thread
+   */
+  @AfterClass
+  def tearDownClass() {
+    verifyNoUnexpectedThreads()
+  }
+
+  /**
+   * Verifies that threads which are known to cause transient failures in subsequent tests
+   * have been shutdown. These include threads which make connections to brokers and may
+   * cause issues when broker ports are reused (e.g. auto-create topics) as well as threads
+   * which reset static JAAS configuration.
+   */
+  def verifyNoUnexpectedThreads() {
+    val unexpectedThreadNames = Set("controller-event-thread",
+                                    "kafka-producer-network-thread",
+                                    "kafka-coordinator-heartbeat-thread",
+                                    "ZkClient-EventThread")
+    def allThreads = Thread.getAllStackTraces.keySet.asScala.map(thread => thread.getName())
+    val (threads, noUnexpected) = TestUtils.computeUntilTrue(allThreads) { threads =>
+      threads.forall(t => unexpectedThreadNames.forall(s => !t.contains(s)))
+    }
+    assertTrue(s"Unexpected threads $threads", noUnexpected)
```

Comment:
I did intend to output all threads (if we are cleaning up properly, there should be very few). I was thinking that if a test failed during development, rather than show just the four unexpected threads, showing all threads would encourage closing anything else that may be left behind too (like an executor for instance). I have updated the message.